[Feature] Automatic tap changer documentation (#606)
PowerGridModel/power-grid-model: docs/user_manual/components.md

#### Input

| name   | data type | unit       | description                                |                 required                  |  update  |            valid values            |
| ------ | --------- | ---------- | ------------------------------------------ | :---------------------------------------: | :------: | :--------------------------------: |

> | `r1`   | `double`  | ohm (Ω)    | positive-sequence serial resistance        |                 &#10004;                  | &#10060; | `r1` and `x1` cannot be both `0.0` |

Reviewer: Why all the changes from "zero" to 0 and 0.0?

Author: because they are all `double` values. I found them confusing when going through the code. I can change it back if you prefer

Reviewer: `0` and `0.0` is fine by me as well. For me personally `zero` was also sufficienct

## Fault

* type name: `fault`
* base: {hoverxreftooltip}`user_manual/components:base`

> `fault` defines a short circuit location in the grid. At the time of writing, a fault can only happen at a `node`.

Reviewer: Not just at the time of writing; For now, a fault can only happen at a node. Maybe just change it in "A fault can only happen at a `node`."

## Regulator

* type name: `regulator`
* base: {hoverxreftooltip}`user_manual/components:base`

`regulator` is an abstract regulator that is coupled to a given `regulated_object`. For each `regulator`, a switch is defined between
the `regulator` and the `regulated_object`. Which object types are supported as `regulated_object` is regulator type-dependent.

#### Input

| name               | data type | unit | description                               | required |  update  |        valid values         |
| ------------------ | --------- | ---- | ----------------------------------------- | :------: | :------: | :-------------------------: |
| `regulated_object` | `int32_t` | -    | ID of the regulated object                | &#10004; | &#10060; | a valid regulated object ID |
| `status`           | `int8_t`  | -    | connection status to the regulated object | &#10004; | &#10004; |         `0` or `1`          |

### Transformer tap regulator

* type name: `transformer_tap_regulator`
* base: {hoverxreftooltip}`user_manual/components:regulator`

`transformer_tap_regulator` defines a regulator for transformers in the grid.

> At the time of writing, a transformer tap regulator regulates a component that is either a {hoverxreftooltip}`user_manual/components:transformer` or a {hoverxreftooltip}`user_manual/components:Three-Winding Transformer`.

Reviewer 2: ```suggestion
A transformer tap regulator regulates a component that is either a {hoverxreftooltip}`user_manual/components:transformer` or a {hoverxreftooltip}`user_manual/components:Three-Winding Transformer`.
```
A lot of things are temporary to mention this everywhere.

Author: the reason is that documentation is very easily outdated (easy to miss something). changed it back to `A` but that was my initial rationale

> The transformer tap regulator overloads the `tap_pos` of the transformer it regulates in the range set by the user via `tap_min` and `tap_max` (i.e., `(tap_min <= tap_pos <= tap_max)` or `(tap_min >= tap_pos >= tap_max)`).

Reviewer 2: While it is most accurate, `overloads` here is a non electrical term users can make do without.
(This is a nitpick, resolve without changes if you wish to)

#### Input

| name                       | data type                                                                                                                                                                                                                                                                                                          | unit     | description                                                                                             |                         required                         |  update  |   valid values    |
| -------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------- | ------------------------------------------------------------------------------------------------------- | :------------------------------------------------------: | :------: | :---------------: |

> | `control_side`             | {py:class}`BranchSide <power_grid_model.enum.BranchSide>` if the regulated object is a {hoverxreftooltip}`user_manual/components:transformer` and {py:class}`Branch3Side <power_grid_model.enum.Branch3Side>` if it the regulated object is a {hoverxreftooltip}`user_manual/components:Three-Winding Transformer` | -        | the controlled side of the transformer                                                                  | &#10024; only for power flow with automatic tap changing | &#10004; |                   |

Reviewer: Why are `u_set` and `u_band` required, but is `control_side` only needed when you want to perform a power flow with automatic tap?

Author: you're right

| `u_set`                    | `double`                                                                                                                                                                                                                                                                                                           | volt (V) | the voltage setpoint (at the center of the band)                                                        |                         &#10004;                         | &#10004; |      `>= 0`       |

> | `u_band`                   | `double`                                                                                                                                                                                                                                                                                                           | volt (V) | the width of the voltage band                                                                           |                         &#10004;                         | &#10004; | `> 0` (see below) |

Author: TODO: elaborate more on lower bound/higher bound vs. per side

| `line_drop_compensation_r` | `double`                                                                                                                                                                                                                                                                                                           | ohm (Ω)  | compensation for voltage drop resistance during transport (see [below](#line-drop-compensation))        |                  &#10060; default `0.0`                   | &#10004; |      `>= 0`       |
| `line_drop_compensation_x` | `double`                                                                                                                                                                                                                                                                                                           | ohm (Ω)  | compensation for voltage drop due to  reactance during transport (see [below](#line-drop-compensation)) |                  &#10060; default `0.0`                   | &#10004; |      `>= 0`       |

The following additional requirements exist on the input parameters.

> - The `control_side` supports {py:class}`BranchSide <power_grid_model.enum.BranchSide>` if it regulates a {hoverxreftooltip}`user_manual/components:transformer` and {py:class}`Branch3Side <power_grid_model.enum.Branch3Side>` if it regulates a {hoverxreftooltip}`user_manual/components:Three-Winding Transformer`.

Reviewer 2: ```suggestion
```
I think the message is clear from the table above.

> - The rated voltage at the `tap_side` must be greater than or equal to the rated voltage at the `control_side`.

Reviewer 2: ```suggestion
- The automatic tap changer algorithm currently only supports tap changers connected at HV side of the transformer. Hence, the rated voltage of node at the `tap_side` must be greater than or equal to the rated voltage of node at the other side of transformer.
```

Control side can be both HV or LV. But only tap_side can be HV.

Author: i resolved this, but semi-related: what's the difference between `u1`/`u2` of the transformers and the `u_rated` of the nodes they are connected to?

Reviewer 2: All these 4 can be different values. `u1`, `u2` are about the actual transformer parameters. They would usually determine the transformation ratio of voltage.

`u_rated` usually is almost equal to `u1` or `u2`. But it does not have to be equal. All the associated `%` or p.u. input grid parameters get affected by the `u_rated` voltages. You can attach any `u1`/`u2` transformer to any `u_rated_1` / `u_rated_2`

While manufacturers could indeed combine all these and consider only 1 value or 2  values, they remain separate as they are all linked to other parameters individually within & outside of power flow calculations.

Also, we could have done ranking against `u1` or `u2`, but that can potentially give more groups of ranks. Different brands of transformer might have different ratings: like maybe some have `10000` or some have `10050` (There might be better examples for this)

**TLDR**: And in relevance of this comment, the check we do is against u_rated of both nodes.

Author: great writeup. thanks!

  $\left|z_{\text{compensation}}\right|\left|I_{\text{node}}\right| \ll \frac{\text{tap_size}/u_\text{tap_side}}{\left(1 + \left(\left|\text{tap_max}\right| - \left|\text{tap_nom}\right|\right)\frac{\text{tap_size}}{\text{u_tap_side}}\right)^2}u_{\text{control_side}}$

These requirements make sure no edge cases with undefined behavior are encountered. Typical real-world power grids already satisfy these requirements and they should therefore not cause any problems.

#### Steady state output

| name      | data type | unit | description          |
| --------- | --------- | ---- | -------------------- |
| `tap_pos` | `int8_t`  | -    | optimal tap position |

#### Short circuit output

A `transformer_tap_regulator` has no short circuit output.

#### Electric Model

The transformer tap regulator itself does not have a direct contribution to the grid state.
Instead, it regulates the tap position of the regulated object until the voltage at the control side is in the specified voltage band:

$$
   \begin{eqnarray}

>       U_{\text{control}}$ is in the range $\left[U_{\text{set}} - \frac{U_{\text{band}}}{2}, U_{\text{set}} + \frac{U_{\text{band}}}{2}\right]

Reviewer: "is in the range" should be as text. Formatting is not good now, see: https://power-grid-model--606.org.readthedocs.build/en/606/user_manual/components.html#id40

   \end{eqnarray}
$$

##### Line drop compensation

The transformer tap regulator tries to regulate the voltage in a specified virtual location in the grid, according to the folowing model.

```txt
tap_side   control_side                        part of grid where voltage is to be regulated
------^\oo -*---------------virtual_impedance----*
      |    U_node, I_node         Z_comp         U_control
      |                                          |
     regulator <=================================/
```

The control voltage is the voltage at the node, compensated with the voltage drop corresponding to the specified line drop compensation.

$$
   \begin{eqnarray}
      & Z_{\text{compensation}} = r_{\text{compensation}} + \mathrm{j} x_{\text{compensation}} \\
      & U_{\text{control}} = \left|\underline{U}_{\text{node}} - \underline{I}_{\text{node}} \cdot \underline{Z}_{\text{compensation}}\right|

>    \end{eqnarray}

Reviewer 2: `u_control_side_node + i_out_transformer * z`
A node does not have a current. 

or maybe as is `u_control_side_node - i_transformer * z` with a repeat refence about the direction of current. The direction part always confuses the user at first.

Author: whoops, you're right.. fixed

Author: fixed again cause of the sign conventions: it's canceling minus signs

`u_control_side_node - i_out_transformer * z = u_control_side_node + i_transformer * z`
